src/main/jbake/content/blog/2017/03/17/apm-adjusting-sampling-rate.adoc

```diff
@@ -0,0 +1,385 @@
+to limit the amount of data being collected or stored. The techniques for that
+are diverse, ranging from a simple "percentage based sampling" of the
+incoming requests, up to complex heuristics to keep/discard traces based on the
+code path.
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-03-17-teaser.png[Final result]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-03-17-teaser.png[Final result]
+endif::[]
+
+It's also a good practice to empower the admin, or an automated monitoring system,
+with the capability of adjusting the sampling based on the current conditions.
+For instance, we'd probably want to get more traces from a
+link:/blog/2017/02/13/monitoring-canary-deployments.html[newly deployed version]
+of an existing application. We might want to compare the performance and adjust
+the percentage in steps. Similarly, if we are experiencing traffic peaks and need
+every resource available, we might enable tracing only to a small percentage of
+the incoming requests.
+
+For the simple scenario, let's call it "percentage based sampling", we have a few
+alternatives on how to implement it:
+
+=== Application-based decision
+
+In this scenario, we embed the logic into the application, using the percentage
+as the probability a random number is within a range. In other words: 20% sampling
+would be 20% chance that a number is lower or equal to 2 on a range of 1-10.
+Anything more complex than that would be "too much" for our scenario.
+
+As hinted before, we want to externalize the percentage so that an external
+actor can change this. One solution is to store this percentage number into an
+environment variable on the Deployment Configuration (DC). This will, however,
+rollout a new deployment after every change, which might not be desirable in some
+scenarios.
+
+Another solution is to externalize this into a Configuration Map (ConfigMap),
+but then, we also need to let our application listen to events related to this
+ConfigMap. For a microservice, we might end up with a bunch of unrelated
+dependencies, just to decide whether or not to trace a request.
+
+=== OpenShift-based decision
+
+In this scenario, we delegate the decision to our runtime environment. Similar
+to a simple A/B testing scenario, our deployment architecture would allow for
+quick changes in the routing decision. It does require a simple change to our
+application, but it's a small compromise for the benefits it provides.
+
+For that, we'll use one single Image Stream (`is`) with our code, two Deployment
+Configurations (`dc`), two Services (`svc`) and one Route object. Each one of our
+`dc` objects would have a different value for an environment variable called
+`TRACER_ENABLED`. Each `svc` has its own `dc`, and the router sends a
+percentage of the requests to each of the `svc` instances. The main advantage of
+this approach is that the application image is the same for both `tracing-enabled`
+and `tracing-disabled` scenarios.
```

Comment:
Can you add an image with some boxes where you show the logical hierarchy that you describe?

Make explicit why the router is sending a percentage to each svc -- is that the 20% from above?